### PR TITLE
feat(FR-3665): name the dev server after its issue, PR and what it does

### LINKED ISSUE #9045: PORTLESS_APP_NAME skips the FR normalization, so a dev server comes up as fr1234.localhost instead of fr-1234.localhost
`scripts/dev.mjs` resolves the portless app name from two sources, and only one of them normalizes an `FR####` token to `fr-####`.

The branch-derived path (priority 2) is correct:

```js
issueMatch[1].toLowerCase().replace(/^fr/, 'fr-').replace(/-{2,}/g, '-')
```

The `PORTLESS_APP_NAME` path (priority 1, which _wins_ over the branch) only slugifies:

```js
raw.toString().toLowerCase().replace(/[^a-z0-9-]+/g, '-')…
```

So the same issue number produces different hostnames depending on which path supplied it:

|input|branch path|PORTLESS_APP_NAME path|
|---|---|---|
|`FR1234`|`fr-1234`|`fr1234`  <-- wrong|
|`FR-1234`|`fr-1234`|`fr-1234`|

This is the path the `dev-server` skill uses — it passes the Claude session name as `PORTLESS_APP_NAME` — so a session named after an issue without the dash yields `https://fr1234.localhost:1357` instead of `https://fr-1234.localhost:1357`.

## Fix

Apply one normalization rule in one place: `fr` immediately followed by digits, at a slug boundary, becomes `fr-<digits>`. Both paths then agree, and `frame123`-style names are left alone.

Extract the derivation into a small importable module so it can be unit tested — `dev.mjs` is a top-level side-effecting script (it spawns the portless daemon on import) and cannot be imported from a test as-is.

JIRA Issue: FR-3665

### PR DESCRIPTION
**Pull request by yomybaby**
Resolves #9045 (FR-3665)

The Portless app name — the dev URL's subdomain — is resolved from two sources in `scripts/dev.mjs`, and only one of them normalized an `FR####` token to `fr-####`.

The branch-derived path (priority 2) did it:

```js
issueMatch[1].toLowerCase().replace(/^fr/, 'fr-').replace(/-{2,}/g, '-')
```

The `PORTLESS_APP_NAME` path (priority 1, which **wins** over the branch) only slugified:

```js
raw.toString().toLowerCase().replace(/[^a-z0-9-]+/g, '-')…
```

So the same issue produced a different hostname depending on which side supplied the name:

| input | branch path | `PORTLESS_APP_NAME` path |
|---|---|---|
| `FR1234` | `fr-1234` | **`fr1234`** ← wrong |
| `FR-1234` | `fr-1234` | `fr-1234` |

That is the path the `dev-server` skill uses — it passes the Claude Code session name as `PORTLESS_APP_NAME` — so a session named after an issue without the dash came up at `https://fr1234.localhost` instead of `https://fr-1234.localhost`.

## Fix

One rule, one place. Both paths now go through `sanitizeAppName`, which normalizes `fr` + digits **at a slug boundary**:

```js
.replace(/(^|-)fr(\d+)(?=-|$)/g, '$1fr-$2')
```

The boundary anchor is what keeps names that merely start with "fr" intact — `frame123` and `french75` are untouched.

The derivation moves into `scripts/portless-app-name.mjs` so it can be unit tested: `dev.mjs` starts the Portless daemon at import time, so a test cannot import it as-is. `dev.mjs` keeps the same priority order and produces the same name for every input it produced correctly before.

## Naming the dev server after what it is

Second commit on this branch, and the reason the PR title now says `feat`. The dev URL was the issue key and nothing else, so a wall of tabs read `fr-3654`, `fr-3652`, `fr-3475` with no way to tell them apart without opening them. It now composes three parts:

```
https://fr-3665-pr9049-statusline.localhost:1357
        \_____/ \____/ \________/
        branch  looked  PORTLESS_APP_NAME
        issue   up      (the only part a caller supplies)
```

Identifiers first — they are what you scan for and they are a fixed width — and the descriptive part last, so the 50-char cap truncates the characters a reader can most easily guess at.

**The descriptive part prefers a human-supplied name over the PR title.** Slugifying a title lands anywhere between good and useless, and nobody can fix a bad one:

| PR | derived from title | |
|---|---|---|
| #9050 | `drop-portless-placeholder` | good |
| #9048 | `redesign-scheduling-history` | good |
| #9049 | `normalize-fr-everywhere` | thin |
| #9038 | `update-astryx-0` | useless ("update Astryx to 0.5.0") |

`/rename` already exists and the dev-server skill already plumbs it through, so `PORTLESS_APP_NAME` now supplies **only that word** and `dev.mjs` adds the identifiers — a caller cannot forget them, and either identifier is stripped back out if one is passed anyway rather than repeated.

Branch names in this repo are mostly bare (`FR-3314`, `FR-3494`), so the title lookup is what makes the automatic path useful at all. One `gh pr view` per boot (~0.65 s) returns number and title together. It is allowed to fail — no `gh`, not logged in, no PR yet, offline — and the name just loses its `prNNNN` part; the last good answer is cached so an offline boot keeps the URL it had instead of silently renaming the server out from under an open tab.

Fallbacks, all covered by tests:

| situation | name |
|---|---|
| `/rename` + PR | `fr-3665-pr9049-statusline` |
| no `/rename` | `fr-3665-pr9049-normalize-fr-everywhere` |
| no PR yet | `fr-3665-statusline` |
| no issue in branch | `pr9049-statusline` |
| caller passed `fr-3665` as the word | `fr-3665-pr9049` (not repeated) |
| `PORTLESS_APP_NAME_EXACT=1` | `v26-4-8-rc-3` (verbatim) |
| nothing at all | `null` → `portless run` auto-derives |

**Trade-off worth a reviewer's attention:** opening a PR changes the URL on the next dev restart, which costs one re-login on that origin. That was the accepted cost of having the PR number in the URL; `PORTLESS_APP_NAME_EXACT=1` opts out of composition entirely.

`.claude/skills/dev-server/SKILL.md` is updated in the same commit — its old guidance told the agent to pass the FR number or PR number itself, which would now be stripped.

## Verification

`bash scripts/verify.sh` → `=== ALL PASS ===`

`pnpm vitest run scripts/portless-app-name.test.ts` → **28 passed**, covering the normalization, the boundary cases, both paths agreeing, title-word extraction, identifier de-duplication, truncation that never eats an identifier, and every fallback in the table above.

Every input converges on one issue key (shown with no PR present, so the `prNNNN` part is absent):

| `PORTLESS_APP_NAME` | branch | resolved URL |
|---|---|---|
| `FR1234` | `main` | `https://fr-1234.localhost:1357` |
| `FR-1234` | `main` | `https://fr-1234.localhost:1357` |
| — | `fix/FR-1234-thing` | `https://fr-1234.localhost:1357` |
| — | `FR1234` | `https://fr-1234.localhost:1357` |
| `frame123` | `main` | `https://frame123.localhost:1357` (unchanged) |

**End to end with the real dev server**, not just the helper.

The normalization fix — `PORTLESS_APP_NAME='FR1234' pnpm dev`:

```
$ grep hostname ~/.portless/routes.json | grep 1234
    "hostname": "fr-1234.localhost",       # was fr1234.localhost

$ curl -sk -o /dev/null -w '%{http_code}' https://fr-1234.localhost:1357/
200
```

The composition — `PORTLESS_APP_NAME='statusline' pnpm dev` on this branch (FR-3665, PR #9049):

```
$ grep hostname ~/.portless/routes.json | grep statusline
    "hostname": "fr-3665-pr9049-statusline.localhost",

$ curl -sk -o /dev/null -w '%{http_code}' https://fr-3665-pr9049-statusline.localhost:1357/
200
```

and the statusline picks each of them up as its Portless link.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01GVhEwwuLyb529VUQo2LY67

### CODE REVIEW

File: packages/backend.ai-agent-cli/src/commands/login.ts [hunk 1]
```diff
@@ -36,8 +37,51 @@ export interface LoginData {
...
+    const routes: unknown = JSON.parse(
+      readFileSync(join(stateDir, 'routes.json'), 'utf8'),
+    );
+    if (!Array.isArray(routes)) return undefined;
+    const names = routes
+      .map((route) =>
+        typeof (route as { hostname?: unknown })?.hostname === 'string'
+          ? (route as { hostname: string }).hostname.split('.')[0]
+          : undefined,
+      )
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply packages/backend.ai-agent-cli/src/commands/login.ts]: Agreed, and this was the most valuable of the four — fixed in `3f2e489fb`. Selecting on hostname alone meant a stale `routes.json` entry, or a sibling worktree on the same issue, could win and hand `login` a dead origin.

`liveAppName` now keeps the `pid` and applies the same ownership test the statusline uses: `realpathSync("/proc/<pid>/cwd")` (which doubles as the liveness check, since a dead pid throws), falling back to `lsof -p <pid> -a -d cwd -Fn` where there is no `/proc`. Routes with `pid <= 0` — the alias entries — are dropped, since they have no cwd to match.

Preference order is now: a live route owned by this checkout, then any live route, then the bare issue key. Exact-match-then-shortest still breaks ties within each tier, so a checkout also running `fr-3665-alt` resolves to its main server.

Four tests added for the new cases: dead pid ignored, dead skipped in favour of live, alias route ignored, and a route entry with no `pid` field ignored. 10 tests total on this function, plus the full `agent-cli` suite at 419.

File: scripts/dev.mjs [hunk 2]
```diff
@@ -63,29 +64,55 @@ spawnSync('portless', proxyArgs, { stdio: 'inherit' });
...
+    return null;
+  }
+}
+
+const appName = resolveAppName({
+  envName: process.env.PORTLESS_APP_NAME,
+  branch,
+  pr: lookupPr(branch),
+  exact: !!process.env.PORTLESS_APP_NAME_EXACT?.trim(),
+});
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply scripts/dev.mjs:125]: Right — `pr: lookupPr(branch)` was an argument expression, so it evaluated regardless of `exact`. Fixed in `3f2e489fb` by resolving the flag first:

```js
const exactAppName = !!process.env.PORTLESS_APP_NAME_EXACT?.trim();
const appName = resolveAppName({
  envName: process.env.PORTLESS_APP_NAME,
  branch,
  pr: exactAppName ? null : lookupPr(branch),
  exact: exactAppName,
});
```

Exact mode now costs no `gh` call at all, so the advertised opt-out really opts out.

File: scripts/dev.mjs [hunk 3]
```diff
@@ -63,29 +64,55 @@ spawnSync('portless', proxyArgs, { stdio: 'inherit' });
...
+    { encoding: 'utf8', timeout: 4000, stdio: ['ignore', 'pipe', 'ignore'] },
+  );
+  if (probe.status === 0 && probe.stdout) {
+    try {
+      const pr = JSON.parse(probe.stdout);
+      if (pr?.number) {
+        mkdirSync(dirname(cacheFile), { recursive: true });
+        writeFileSync(cacheFile, JSON.stringify(pr));
+        return pr;
+      }
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply scripts/dev.mjs]: Fixed in `3f2e489fb`, though the failure mode is milder than described — the writes were already inside a `try`, so a cache-write error could not abort the dev server. What it *did* do is worse than nothing in a quiet way: the `catch` fell through to reading the cache, so a fresh, successful `gh` answer got discarded in favour of a stale entry or none at all.

Parsing and persisting are now separate, and persistence is explicitly best-effort:

```js
try { mkdirSync(dirname(cacheFile), { recursive: true }); writeFileSync(cacheFile, JSON.stringify(pr)); }
catch { /* Best-effort; the fresh lookup below is still good. */ }
return pr;
```

File: scripts/portless-app-name.mjs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    .toLowerCase()
+    .replace(/[^a-z0-9-]+/g, '-')
+    .replace(/-{2,}/g, '-')
+    .replace(/^-+|-+$/g, '')
+    .slice(0, MAX_LEN)
+    .replace(/-+$/, '') // the slice can land mid-word and leave a trailing dash
+    // `fr1234` and `fr-1234` name the same issue, so they must produce the same
+    // hostname whichever source supplied the name. Anchoring on a slug boundary
+    // keeps words that merely start with "fr" (frame123) intact.
+    .replace(ISSUE_TOKEN, '$1fr-$2');
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply scripts/portless-app-name.mjs]: Confirmed and fixed in `3f2e489fb`. Reproduced before the change:

```
$ sanitizeAppName("a".repeat(44) + "-fr123")   // 50 chars in
input len 50 -> output len 51   "aaaaaaa-fr-123"
```

`ISSUE_TOKEN` now runs before the slice, with the trailing-dash trim kept after it:

```js
.replace(ISSUE_TOKEN, "$1fr-$2")
.slice(0, MAX_LEN)
.replace(/-+$/, "")
```

50 -> 50, 52 -> 50, 64 -> 50. Added a regression test pinning the boundary inputs at <= MAX_LEN.
